patches/0013-p256-armv8-optimizations

```diff
@@ -2429,20 +2457,460 @@ Index: aws-lc/third_party/boringssl/crypto/fipsmodule/ec/p256-x86_64_test.cc
++  __attribute__((aligned(64))) P256_POINT table[16];
+   for (size_t i = 0; i < 16; i++) {
+     OPENSSL_memset(table[i].X, 3 * i, sizeof(table[i].X));
+     OPENSSL_memset(table[i].Y, 3 * i + 1, sizeof(table[i].Y));
+@@ -71,7 +70,7 @@ TEST(P256_X86_64Test, SelectW5) {
+ 
+ TEST(P256_X86_64Test, SelectW7) {
+   // Fill a table with some garbage input.
+-  alignas(64) P256_POINT_AFFINE table[64];
++  __attribute__((aligned(64))) P256_POINT_AFFINE table[64];
+   for (size_t i = 0; i < 64; i++) {
+     OPENSSL_memset(table[i].X, 2 * i, sizeof(table[i].X));
+     OPENSSL_memset(table[i].Y, 2 * i + 1, sizeof(table[i].Y));
+@@ -99,11 +98,12 @@ TEST(P256_X86_64Test, SelectW7) {
  }
  
-+#if !defined(OPENSSL_AARCH64)
  TEST(P256_X86_64Test, BEEU) {
++#if !defined(OPENSSL_AARCH64)
    if ((OPENSSL_ia32cap_P[1] & (1 << 28)) == 0) {
      // No AVX support; cannot run the BEEU code.
-@@ -165,6 +165,8 @@ TEST(P256_X86_64Test, BEEU) {
+     return;
    }
- }
- 
+-
 +#endif
+   bssl::UniquePtr<EC_GROUP> group(
+       EC_GROUP_new_by_curve_name(NID_X9_62_prime256v1));
+   ASSERT_TRUE(group);
+Index: aws-lc/third_party/boringssl/crypto/fipsmodule/ec/asm/p256_beeu-armv8-asm.pl
+===================================================================
+--- /dev/null
++++ aws-lc/third_party/boringssl/crypto/fipsmodule/ec/asm/p256_beeu-armv8-asm.pl
+@@ -0,0 +1,423 @@
++# Copyright (c) 2020, Amazon Inc.
++#
++# Permission to use, copy, modify, and/or distribute this software for any
++# purpose with or without fee is hereby granted, provided that the above
++# copyright notice and this permission notice appear in all copies.
++#
++# THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
++# WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
++# MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
++# SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
++# WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION
++# OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
++# CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE. */
++#
++# Written by Nevine Ebeid
++# AWS Cryptographic Algorithms Group
++# (nebeid@amazon.com)
++# based on p256_beeu-x86_64-asm.pl (which is based on BN_mod_inverse_odd).
 +
- static bool GetFieldElement(FileTest *t, BN_ULONG out[P256_LIMBS],
-                             const char *name) {
-   std::vector<uint8_t> bytes;
++$flavour = shift;
++while (($output=shift) && ($output!~/\w[\w\-]*\.\w+$/)) {}
++
++$0 =~ m/(.*[\/\\])[^\/\\]+$/; $dir=$1;
++( $xlate="${dir}arm-xlate.pl" and -f $xlate ) or
++( $xlate="${dir}../../../perlasm/arm-xlate.pl" and -f $xlate) or
++die "can't locate arm-xlate.pl";
++
++open OUT,"| \"$^X\" $xlate $flavour $output";
++*STDOUT=*OUT;
++#############################################################################
++# extern int beeu_mod_inverse_vartime(BN_ULONG out[P256_LIMBS],
```

Comment:
I see the assumption below (`n` is odd). But is there somewhere else you can document that this algorithm is only good if `n` is odd (if not documented already). Not many would actually read the asm.

Reply:
It's documented with its declaration in `p256-x86_64.h`

Reply:
Ahh I see here: https://github.com/awslabs/aws-lc/blob/main/crypto/fipsmodule/ec/p256-x86_64.h#L96